File: projects/iffywow/elixir/lib/ithkuil/svg.ex
defmodule Ithkuil.SVG do
  @moduledoc """
  Standalone SVG export with embedded coordinate metadata, and exact tuple
  recovery from any SDK-generated SVG. Mirrors `../../web/src/metadata.js`
  (`sceneToSVG` / `extractCoordinate`).

  Every SVG we generate carries its own coordinate:

    * `data-ithkuil-schema` / `data-ithkuil-integer` attributes
    * a `<metadata id="ithkuil-coordinate">` JSON block whose `&` and `<`
      are escaped as `\\u0026` / `\\u003c` (both only ever occur inside JSON
      strings), so the metadata body can never break strict XML parsing

  so `extract_coordinate(render(model)) == {:ok, model.coordinate}` holds with
  no vision, OCR, or geometric inference. Recognition of *unannotated* glyphs
  is a deferred, separate subsystem — deliberately not implemented here.
  """

  alias Ithkuil.JSON

  @kind_attrs %{
    "base" => ~s(stroke-width="3"),
    "modifier" => ~s(stroke-width="2.4"),
    "diacritic" => ~s(stroke-width="2"),
    "connector" => ~s(stroke-width="1" stroke-dasharray="4 3" opacity="0.65"),
    "socket-marker" => ~s(stroke-width="1.25")
  }

  @doc """
  Render a scene model (from `Ithkuil.Scene.compile/2` / `Ithkuil.to_scene/1`)
  to a standalone SVG document string. `mode` is `:compact` or `:exploded`.
  """
  @spec render(Ithkuil.Scene.model(), :compact | :exploded) :: String.t()
  def render(model, mode \\ :compact)

  def render(model, mode) when mode in [:compact, :exploded] do
    [x, y, w, h] = model.view_box

    body =
      model.nodes
      |> Enum.map(&node_group(&1, mode))
      |> Enum.reject(&(&1 == ""))
      |> Enum.join("\n")

    meta = metadata_json(model, mode)

    integer_attr =
      if truthy(model.integer),
        do: ~s( data-ithkuil-integer="#{esc(model.integer)}"),
        else: ""

    title =
      if truthy(model.latinized),
        do: "  <title>#{esc(model.latinized)}</title>\n",
        else: ""

    ~s(<svg xmlns="http://www.w3.org/2000/svg" viewBox="#{f(x)} #{f(y)} #{f(w)} #{f(h)}") <>
      ~s( data-ithkuil-schema="#{esc(model.schema)}"#{integer_attr}) <>
      ~s( fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round">\n) <>
      title <>
      ~s(  <metadata id="ithkuil-coordinate">#{meta}</metadata>\n) <>
      body <> "\n</svg>\n"
  end

  def render(_model, mode) do
    raise ArgumentError, "unknown mode: #{inspect(mode)}"
  end

  @doc """
  Exact coordinate recovery from an SDK-generated SVG string. Returns
  `{:ok, wire}` (the coordinate wire form as Elixir data) or
  `{:error, :no_metadata}` when the metadata was stripped — at which point the
  file is an ordinary unannotated drawing and belongs to the deferred
  recognition subsystem.
  """
  @spec extract_coordinate(String.t()) :: {:ok, Ithkuil.Coord.wire()} | {:error, :no_metadata}
  def extract_coordinate(svg_text) when is_binary(svg_text) do
    case extract_metadata(svg_text) do
      {:ok, %{"coordinate" => coordinate}} when is_list(coordinate) -> {:ok, coordinate}
      _ -> {:error, :no_metadata}
    end
  end

  @doc """
  Full metadata block (`schema`, `latinized`, `integer`, `mode`, `coordinate`)
  as a map with string keys, or `{:error, :no_metadata}`.
  """
  @spec extract_metadata(String.t()) :: {:ok, map()} | {:error, :no_metadata}
  def extract_metadata(svg_text) when is_binary(svg_text) do
    with {:ok, raw} <- find_metadata(svg_text),
         {:ok, parsed} when is_map(parsed) <- JSON.parse(raw) do
      {:ok, parsed}
    else
      _ -> {:error, :no_metadata}
    end
  end

  # ------------------------------------------------------------------
  # Internals
  # ------------------------------------------------------------------

  defp node_group(node, mode) do
    p = Map.fetch!(node, mode)
    path = Map.get(node, :path)

    if p.visible and is_binary(path) and path != "" do
      [tx, ty] = p.translate
      t = "translate(#{f(tx)} #{f(ty)}) rotate(#{f(p.rotate)}) scale(#{f(p.scale)})"
      extra = Map.get(@kind_attrs, node.kind, "")

      ~s(  <g data-node-id="#{esc(node.id)}" data-kind="#{esc(node.kind)}" transform="#{t}" #{extra}) <>
        ~s(><path d="#{esc(path)}"/></g>)
    else
      ""
    end
  end

  # JSON.stringify({schema, latinized, integer, mode, coordinate}) with & and <
  # escaped as JSON \u escapes — identical to web/src/metadata.js.
  defp metadata_json(model, mode) do
    JSON.encode(
      {:object,
       [
         {"schema", model.schema},
         {"latinized", model.latinized},
         {"integer", model.integer},
         {"mode", Atom.to_string(mode)},
         {"coordinate", model.coordinate}
       ]}
    )
    |> String.replace("&", "\\u0026")
    |> String.replace("<", "\\u003c")
  end

  defp find_metadata(svg_text) do
    case Regex.run(~r/<metadata id="ithkuil-coordinate">(.*?)<\/metadata>/s, svg_text) do
      [_, inner] ->
        {:ok, inner}

      nil ->
        case Regex.run(~r/<metadata\b[^>]*>(.*?)<\/metadata>/s, svg_text) do
          [_, inner] -> {:ok, inner}
          nil -> {:error, :no_metadata}
        end
    end
  end

  # JS truthiness for the optional string facts: nil and "" are falsy.
  defp truthy(nil), do: false
  defp truthy(""), do: false
  defp truthy(_), do: true

  defp esc(s) do
    s
    |> to_string()
    |> String.replace("&", "&amp;")
    |> String.replace("<", "&lt;")
    |> String.replace(">", "&gt;")
    |> String.replace("\"", "&quot;")
  end

  defp f(n), do: JSON.format_number(n)
end

File: projects/iffywow/python/ithkuil/svg.py
"""Standalone SVG export with embedded coordinate metadata + exact recovery.

Byte-compatible port of ``web/src/metadata.js`` ``sceneToSVG`` /
``extractCoordinate``: same attribute set, same ``<metadata
id="ithkuil-coordinate">`` JSON block, so

    extract_coordinate(render(compile_scene(c))) == word_to_wire(c)

holds with no vision, OCR, or geometric inference.  Recognition of
UNANNOTATED glyphs is a deferred, separate subsystem -- deliberately not
implemented here.

Byte-parity notes:

* The metadata JSON uses ``json.dumps(..., separators=(",", ":"),
  ensure_ascii=False)`` and the fixed key order ``schema, latinized,
  integer, mode, coordinate`` to match JS ``JSON.stringify`` exactly
  (compact separators; insertion-ordered keys; non-ASCII text unescaped).
* ``&`` and ``<`` are then escaped as ``\\u0026`` / ``\\u003c``.  Both only
  ever occur inside JSON strings, so the escaped body is still valid JSON
  AND can never break strict XML parsing.
"""

from __future__ import annotations

import json
import xml.etree.ElementTree as ET
from typing import Any, Dict, Optional

from .scene import js_num as _num

KIND_ATTRS = {
    "base": 'stroke-width="3"',
    "modifier": 'stroke-width="2.4"',
    "diacritic": 'stroke-width="2"',
    "connector": 'stroke-width="1" stroke-dasharray="4 3" opacity="0.65"',
    "socket-marker": 'stroke-width="1.25"',
}


def _esc(s: Any) -> str:
    """XML attribute/text escaping, replacement order matching metadata.js
    (& first so later entities are not double-escaped)."""
    return (
        str(s)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
    )


def render(model: Dict[str, Any], mode: str = "compact") -> str:
    """Render a scene model (from :func:`ithkuil.scene.compile_scene` or the
    Elixir codec) to a standalone SVG document string."""
    if mode not in ("compact", "exploded"):
        raise ValueError(f"unknown mode: {mode}")

    x, y, w, h = model["viewBox"]

    lines = []
    for node in model["nodes"]:
        p = node.get(mode)
        if not p or not p.get("visible") or not node.get("path"):
            continue
        t = (
            f"translate({_num(p['translate'][0])} {_num(p['translate'][1])})"
            f" rotate({_num(p['rotate'])}) scale({_num(p['scale'])})"
        )
        extra = KIND_ATTRS.get(node["kind"], "")
        lines.append(
            f'  <g data-node-id="{_esc(node["id"])}" data-kind="{_esc(node["kind"])}"'
            f' transform="{t}" {extra}><path d="{_esc(node["path"])}"/></g>'
        )
    body = "\n".join(lines)

    # Escape & and < as JSON \u escapes (both only ever occur inside JSON
    # strings) so the metadata body can never break strict XML parsing.
    meta = json.dumps(
        {
            "schema": model["schema"],
            "latinized": model.get("latinized"),
            "integer": model.get("integer"),
            "mode": mode,
            "coordinate": model["coordinate"],
        },
        separators=(",", ":"),
        ensure_ascii=False,
    ).replace("&", "\\u0026").replace("<", "\\u003c")

    integer = model.get("integer")
    integer_attr = f' data-ithkuil-integer="{_esc(integer)}"' if integer else ""
    latinized = model.get("latinized")
    title = f"  <title>{_esc(latinized)}</title>\n" if latinized else ""

    return (
        f'<svg xmlns="http://www.w3.org/2000/svg" viewBox="{_num(x)} {_num(y)} {_num(w)} {_num(h)}"'
        f' data-ithkuil-schema="{_esc(model["schema"])}"{integer_attr}'
        f' fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round">\n'
        f"{title}"
        f'  <metadata id="ithkuil-coordinate">{meta}</metadata>\n'
        f"{body}\n</svg>\n"
    )


def extract_metadata(svg_text: str) -> Optional[Dict[str, Any]]:
    """Full embedded metadata block ``{schema, latinized, integer, mode,
    coordinate}`` or ``None`` when absent/unparseable (mirrors JS
    ``extractMetadata``: any failure degrades to None, never raises)."""
    try:
        root = ET.fromstring(svg_text)
    except (ET.ParseError, ValueError, TypeError):
        return None
    chosen = None
    for elem in root.iter():
        if elem.tag.rpartition("}")[2] == "metadata":
            if elem.get("id") == "ithkuil-coordinate":
                chosen = elem
                break
            if chosen is None:
                chosen = elem  # fallback: first <metadata>
    if chosen is None or not chosen.text:
        return None
    try:
        parsed = json.loads(chosen.text)
    except json.JSONDecodeError:
        return None
    return parsed if isinstance(parsed, dict) else None


def extract_coordinate(svg_text: str) -> Optional[list]:
    """Exact tuple recovery from an SDK-generated SVG: the coordinate wire
    form, or ``None`` when the metadata was stripped (at which point the
    file is an ordinary unannotated drawing and belongs to the deferred
    recognition subsystem)."""
    meta = extract_metadata(svg_text)
    if isinstance(meta, dict) and isinstance(meta.get("coordinate"), list):
        return meta["coordinate"]
    return None
